File: Top_Interview_Questions/generateParenthesis.go
func generateParenthesis(n int) []string {
    return backtrack([]string{}, "", n, 0);
}

func backtrack(result []string, cur_s string, n int, m int) []string {
    if n == 0 && m == 0 {
        result = append(result, cur_s);
        return result;
    }

    if n > 0 {
        result = backtrack(result, cur_s + "(", n-1, m+1);
    }

    if m > 0 {
        result = backtrack(result, cur_s + ")", n, m-1);
    }
    
    return result;
}

File: NeetCode_Roadmap/generate_parenthesis.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        result = []
        def backtrack(cur_s: str, rem: int, unclosed: int):
            if rem + unclosed == 0:
                result.append(cur_s)
                return

            if rem > 0:
                backtrack(cur_s + "(", rem-1, unclosed+1)
            
            if unclosed > 0:
                backtrack(cur_s+")", rem, unclosed-1)

        backtrack("", n, 0)
        return result